File: game-engine/src/main/java/org/luxons/sevenwonders/game/resources/ResourceTransactions.java
package org.luxons.sevenwonders.game.resources;

import java.util.Collection;
import java.util.HashMap;
import java.util.Map;
import java.util.Objects;
import java.util.Set;
import java.util.stream.Collectors;

import org.luxons.sevenwonders.game.api.Table;

public class ResourceTransactions {

    private final Map<Provider, Resources> resourcesByProvider;

    public ResourceTransactions() {
        this.resourcesByProvider = new HashMap<>();
    }

    public ResourceTransactions(Collection<ResourceTransaction> transactions) {
        this();
        transactions.forEach(t -> add(t.getProvider(), t.getResources()));
    }

    public void add(Provider provider, Resources resources) {
        resourcesByProvider.putIfAbsent(provider, new Resources());
        resourcesByProvider.merge(provider, resources, Resources::addAll);
    }

    public void remove(Provider provider, Resources resources) {
        resourcesByProvider.compute(provider, (p, prevResources) -> {
            if (prevResources == null) {
                throw new IllegalStateException("Cannot remove resources from resource transactions");
            }
            return prevResources.minus(resources);
        });
    }

    public void execute(Table table, int playerIndex) {
        toTransactions().forEach(t -> t.execute(table, playerIndex));
    }

    public Set<ResourceTransaction> toTransactions() {
        return resourcesByProvider.entrySet()
                                  .stream()
                                  .filter(e -> !e.getValue().isEmpty())
                                  .map(e -> new ResourceTransaction(e.getKey(), e.getValue()))
                                  .collect(Collectors.toSet());
    }

    public Resources asResources() {
        return resourcesByProvider.values().stream().reduce(new Resources(), Resources::addAll);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        ResourceTransactions that = (ResourceTransactions) o;
        return Objects.equals(resourcesByProvider, that.resourcesByProvider);
    }

    @Override
    public int hashCode() {
        return Objects.hash(resourcesByProvider);
    }
}

File: game-engine/src/main/kotlin/org/luxons/sevenwonders/game/resources/ResourceTransactions.kt
package org.luxons.sevenwonders.game.resources

import org.luxons.sevenwonders.game.Player
import org.luxons.sevenwonders.game.api.ApiCountedResource
import org.luxons.sevenwonders.game.api.toCountedResourcesList

fun Map<Provider, Resources>.toTransactions(): ResourceTransactions =
    filterValues { !it.isEmpty() }
        .map { (p, res) -> ResourceTransaction(p, res.toCountedResourcesList()) }
        .toSet()

fun ResourceTransactions.asResources(): Resources = flatMap { it.resources }.asResources()

fun ResourceTransaction.asResources(): Resources = resources.asResources()

fun List<ApiCountedResource>.asResources(): Resources = map { it.asResources() }.merge()

fun ApiCountedResource.asResources(): Resources = resourcesOf(type to count)

internal fun ResourceTransactions.execute(player: Player) = forEach { it.execute(player) }

internal fun ResourceTransaction.execute(player: Player) {
    val board = player.board
    val price = board.tradingRules.computeCost(this)
    board.removeGold(price)
    val providerPosition = provider.boardPosition
    val providerBoard = player.getBoard(providerPosition)
    providerBoard.addGold(price)
}
